File: Python/272__Closest_Binary_Search_Tree_Value_II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque

class Solution(object):
    def closestKValues(self, root, target, k):
        """
        :type root: TreeNode
        :type target: float
        :type k: int
        :rtype: List[int]
        """
        def inorder(root):
            if not root:
                return
            inorder(root.left)
            if len(res) < k:
                res.append(root.val)
            elif abs(res[0] - target) > abs(root.val - target):
                res.popleft()
                res.append(root.val)
            else:
                return
            inorder(root.right)

        res = deque([])
        inorder(root)
        return list(res)

File: Java/272__Closest_Binary_Search_Tree_Value_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> closestKValues(TreeNode root, double target, int k) {
        List<Integer> result = new LinkedList();
        Stack<TreeNode> stack = new Stack();
        do {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                root = stack.pop();
                if (result.size() < k)
                    result.add(root.val);
                else {
                    if (Math.abs(result.get(0) - target) >= Math.abs(root.val - target)) {
                        result.remove(0);
                        result.add(root.val);
                    } else
                        return result;
                }
                root = root.right;
            }
        } while (!stack.isEmpty() || root != null);

        return result;
    }
}
